webapp/channels/src/workers/sw.js

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// Update this every time code is changed.
+// This is needed to delete old cache entries.
+const cacheName = 'v1';
+const ROOT_HTML_REQUEST = new Request('root.html');
+
+const putInCache = async (request, response) => {
+    const cache = await caches.open(cacheName);
+    if (request.destination === 'document') {
+        await cache.put(ROOT_HTML_REQUEST, response);
+    } else {
+        await cache.put(request, response);
+    }
+};
+
+const handleCacheRequest = async (request) => {
+    // Using the stale-while-revalidate approach for cached requests to keep them in sync.
+    let responseFromCache;
+    if (request.destination === 'document') {
+        // If it's an HTML document, then we check for a separate key.
+        responseFromCache = await caches.match(ROOT_HTML_REQUEST);
+    } else {
+        responseFromCache = await caches.match(request);
+    }
+    if (responseFromCache) {
+        // Non-blocking cache insertion.
+        fetchFromNetwork(request);
```

Comment:
What happens if this returns an updated version of something? Is there some way to alert the web app that there's a fresher version of this data that can be loaded by refreshing?

Reply:
We could have some complicated way. But mostly I have relied on the fact that the user can simply reload to get the latest updated content from the cache.

Reply:
How does this behave with the "refresh to get a new version?" bar that we already have though? It seems like this would make it so that clicking on the bar would refresh the page with the new version cached in the background, so they'd have to refresh twice to get the updated version of the app